### PR TITLE
Adds new features to ClassList

### PR DESCRIPTION
**Pull request by DrPaulSharp**
We add the ability to:
 - set an item in the ClassList using an object
 - determine the class handle by ensuring "isinstance" is true for all elements (allows ProtectedParameters to be included correctly.)
 - iadd and extend the ClassList with the expected behaviour for individual objects.

### CODE REVIEW

File: RAT/classlist.py [hunk 1]
```diff
@@ -61,15 +61,22 @@ def __repr__(self):
...
-        """Assign the values of an existing object's attributes using a dictionary."""
-        self._setitem(index, set_dict)
+    def __setitem__(self, index: int, item: Union['RAT.models', dict[str, Any]]) -> None:
+        """Assign the values of an existing object's attributes using either a replacement object or a dictionary
+        containing key-value pairs.
+        """
+        self._setitem(index, item)
 
-    def _setitem(self, index: int, set_dict: dict[str, Any]) -> None:
+    def _setitem(self, index: int, item: Union['RAT.models', dict[str, Any]]) -> None:
```
Comment (StephenNneji) [inline RAT/classlist.py:70]: I think it might be better to support one type i.e only models or dict (my pick is models) then provide a fallback for anyone who wants to use the alternate e.g. add_rows_as_type. Otherwise there will be failing edge cases in the current design

File: tests/test_classlist.py [hunk 2]
```diff
@@ -563,3 +611,17 @@ def test__get_item_from_name_field(two_name_class_list: 'ClassList',
...
+
+
+@pytest.mark.parametrize("input_list", [
+    ([InputAttributes(name='Alice')]),
+    ([InputAttributes(name='Alice'), SubInputAttributes(name='Bob')]),
+    ([SubInputAttributes(name='Alice'), InputAttributes(name='Bob')]),
+    ([SubInputAttributes(name='Alice'), SubInputAttributes(name='Bob'), InputAttributes(name='Eve')]),
+    ([InputAttributes(name='Alice'), dict(name='Bob')]),
+])
+def test_determine_class_handle(input_list: 'ClassList') -> None:
```
[the remark this answers is not in the document]
Comment (DrPaulSharp) [reply tests/test_classlist.py:623]: Yes, they're valid and would give the results `SubInputAttributes` and `dict` respectively. I'll write a better test.

[the remark this answers is not in the document]
Comment (DrPaulSharp) [reply tests/test_classlist.py:623]: (The first is a problem for the ClassList, but that is picked up elsewhere)

File: RAT/classlist.py (hunk 1 above)
Comment (DrPaulSharp) [reply RAT/classlist.py:70]: Yeah, that's fine with me. I'll get on it!

**Review by StephenNneji — APPROVED** [review]
